Jinchao fix default role as Volunteer after role deletion

suigunfjc opened the pull request:
# Description
Fix bug:
(PRIORITY LOW) Jae/Raul: After creating a role and setting it to a user, if this role is deleted, the user role becomes Administrator by default. Should be 'Volunteer'.

## Mainly changes explained:
Extend `rolesController.deleteRoleById` :
After role deletion, update all users which currently in that role to volunteer.
Update the user cache after changing users' roles.

## How to test:

1. Log in as an owner user (if you don't have one, use `suigunfjc@gmail.com, Ff12345678!`)
2. Go to dashboard -> Other Link -> permissions management -> add new role
3. Create a new role and assign it to any user (create a new user or change a user's role on its profile page)
4. Permissions management -> click on the new role  -> delete role
5. Go back and check if the users' roles have changed to volunteer on their profile page.

## Note:
The roles displayed on the user management page will be outdated after role deletion or changing from user profile page. Please confirm the user's role by checking the profile page
![Screenshot 2023-03-01 210714](https://user-images.githubusercontent.com/94319381/222336364-ffb983aa-6057-4c82-a6cf-76f080e144f6.png)

vitorsaa2k reviewed (approved):
Hey @suigunfjc, great PR. As you can see in the video it works as intended.

https://user-images.githubusercontent.com/17365161/222520151-3531634f-37cf-450f-9efa-1fff888b9645.mp4

angelinatr15 commented:
Hi @suigunfjc, 

I did not have an owner user so I used the login you provided. I tested this branch by creating a newTest role then I created a new user and assigned it the newTest role. When I checked the profile it does display the intended newTest role for the new user. Lastly I went back into the permission management and deleted the newTest role. When I went back into the new user profile I saw that the role status has changed into volunteer status as the default. (see video below)

I also logged into the new user and checked out the profile everything seemed fine and the status was at volunteer. 

I tested this out once more by creating 2 new user and assigned it a new role I created. I repeated the process and deleted the role. Once I did that both the roles status defaulted to volunteer. 

This PR worked as intended. I did notice that when I went to user management and checked on the new users the deleted role was still displayed.This was also to be expected. 

https://user-images.githubusercontent.com/97488347/222596780-c78578f6-27f4-4008-84b1-6b22bc343741.mov

Lunaliar commented:
Hey, @suigunfjc I also wasn't able to make an owner user so I used the one you provided. I tested this out with two users one an Administrator and another that was a volunteer. When I assigned them my test role and then deleted it they each reverted to administrator roles. I retested after switching their roles to volunteer and testing through the process again to the same outcome. 
Original roles: 
![calo admin](https://user-images.githubusercontent.com/98275089/222613577-aad11b1a-c3b5-4c33-a80f-2a339243f48e.png)
![Jean Volunteer](https://user-images.githubusercontent.com/98275089/222613589-af92ca35-2fe2-49b1-bedc-d63dc3bd96d9.png)
Test Roles: 
![Calo test](https://user-images.githubusercontent.com/98275089/222613664-0abbfb71-18ef-40ee-a322-ace40f8f6f8d.png)
![Jean test](https://user-images.githubusercontent.com/98275089/222613644-c6942dc4-cb4e-4614-8cf7-f718116f29dd.png)
Post-deleted test role:
![calo admin](https://user-images.githubusercontent.com/98275089/222613764-4ed37ede-6f86-4b3e-bd8d-f23962098ae5.png)
![Jean admin](https://user-images.githubusercontent.com/98275089/222613765-5750ea33-88f6-41bc-a417-15ca67b8ca7a.png)

NyahJ reviewed (changes requested):
Hi @suigunfjc,

Most of the function performed well, the added and deleted new roles would affect users selection range in their profile page. However, there is a small issue about the role names:
I used four test roles: 1. TestWithNumber 2.test role 1 (name with number and black space) 3. Test03/02/2023 (name with numbers and "/") 4. TestWith!? (name with "?" and "!")
The test result for names with characters and numbers only worked perfectly, both of adding and deleting. **But after we add the role names with punctuation successfully, we were not able to delete them, because the webpage appeared error/blank (result was different by different marks) so that we could not make it to delete those kinds of roles.**
I have attached a video for your reference, could you pls take a look at it? Thanks!

https://user-images.githubusercontent.com/77643813/222615366-bb644034-51c6-47da-b447-382889eccccf.mp4

suigunfjc commented:
> Hi @suigunfjc,
> 
> Most of the function performed well, the added and deleted new roles would affect users selection range in their profile page. However, there is a small issue about the role names: I used four test roles: 1. TestWithNumber 2.test role 1 (name with number and black space) 3. Test03/02/2023 (name with numbers and "/") 4. TestWith!? (name with "?" and "!") The test result for names with characters and numbers only worked perfectly, both of adding and deleting. But after we add the role names with punctuation successfully, we were not able to delete them, because the webpage appeared error/blank (result was different by different marks) so that we could not make it to delete those kinds of roles. I have attached a video for your reference, could you pls take a look at it? Thanks!

Lucas-E reviewed (approved):
Hey Jinchao, how are you doing? I've tested your PR using the specified user, and everything worked as intended. The role was outdated as expected, but the users roles became volunteer after deleting the created role. Greate job!
Before deleting the role:
![Captura de tela de 2023-03-02 23-19-38](https://user-images.githubusercontent.com/92558269/222615863-5675d388-b1d4-47fb-8bab-7cd5b2419462.png)
![Captura de tela de 2023-03-02 23-21-04](https://user-images.githubusercontent.com/92558269/222615862-2cadf62b-21c1-4792-9ffa-a2055c092826.png)
After deleting the role:
![Captura de tela de 2023-03-02 23-22-38](https://user-images.githubusercontent.com/92558269/222615860-bc247c93-126b-408f-932f-4c82b14e0a72.png)

![Captura de tela de 2023-03-02 23-22-46](https://user-images.githubusercontent.com/92558269/222615859-0a360345-5188-46db-98e6-55ee1ecca412.png)
![Captura de tela de 2023-03-02 23-22-56](https://user-images.githubusercontent.com/92558269/222615858-806e6f2a-609f-4caa-bfd5-582cb023e73e.png)

anant2047 commented:
Hey Jinchao,

I tested the code and it works as intended.

I created a new role - 
<img width="1440" alt="Screen Shot 2023-03-02 at 6 28 49 PM" src="https://user-images.githubusercontent.com/97596031/222616790-ea8b13da-e9fe-48c1-b2d9-01241480a2de.png">


Then assigned to test user - 
<img width="1389" alt="Screen Shot 2023-03-02 at 6 29 31 PM" src="https://user-images.githubusercontent.com/97596031/222616807-1ba305e2-9b26-41ff-a95d-8c752937c780.png">


Then I deleted the role -
<img width="904" alt="Screen Shot 2023-03-02 at 6 29 39 PM" src="https://user-images.githubusercontent.com/97596031/222616827-e4af18b3-2ebb-4bd6-a808-e6af1e56cc8d.png">

I see volunteer role now after deletion for test user - 
<img width="1333" alt="Screen Shot 2023-03-02 at 6 30 05 PM" src="https://user-images.githubusercontent.com/97596031/222616874-d1f029b6-7421-4769-ad5b-f5982326f8b6.png">

suigunfjc commented:
> Hi @suigunfjc,
> 
> Most of the function performed well, the added and deleted new roles would affect users selection range in their profile page. However, there is a small issue about the role names: I used four test roles: 1. TestWithNumber 2.test role 1 (name with number and black space) 3. Test03/02/2023 (name with numbers and "/") 4. TestWith!? (name with "?" and "!") The test result for names with characters and numbers only worked perfectly, both of adding and deleting. **But after we added the role names with punctuation successfully, we were not able to delete them, because the webpage appeared error/blank (result was different by different marks) so that we could not make it to delete those kinds of roles.** I have attached a video for your reference, could you pls take a look at it? Thanks!

Thank you @NyahJ, It is definitely a new bug and I will take a look at it. Though I think it's out of this PR's scope, we can merge this PR first and raise another PR to solve it.

suigunfjc commented:
> Hey, @suigunfjc I also wasn't able to make an owner user so I used the one you provided. I tested this out with two users one an Administrator and another that was a volunteer. When I assigned them my test role and then deleted it they each reverted to administrator roles. I retested after switching their roles to volunteer and testing through the process again to the same outcome. Original roles: ![calo admin](https://user-images.githubusercontent.com/98275089/222613577-aad11b1a-c3b5-4c33-a80f-2a339243f48e.png) ![Jean Volunteer](https://user-images.githubusercontent.com/98275089/222613589-af92ca35-2fe2-49b1-bedc-d63dc3bd96d9.png) Test Roles: ![Calo test](https://user-images.githubusercontent.com/98275089/222613664-0abbfb71-18ef-40ee-a322-ace40f8f6f8d.png) ![Jean test](https://user-images.githubusercontent.com/98275089/222613644-c6942dc4-cb4e-4614-8cf7-f718116f29dd.png) Post-deleted test role: ![calo admin](https://user-images.githubusercontent.com/98275089/222613764-4ed37ede-6f86-4b3e-bd8d-f23962098ae5.png) ![Jean admin](https://user-images.githubusercontent.com/98275089/222613765-5750ea33-88f6-41bc-a417-15ca67b8ca7a.png)

Hi @Lunaliar, thank you for testing this PR. However, I couldn't reproduce your situation, everything goes well on my local host. Could you post a video showing how this issue happened? Thanks!

dkamghem commented:
I reviewed this PR, it works fine. I created a role "Daniel's role" and edit a user to that role. Once the role it deleted, the user goes back to its initial role before the new role. In my specific example, the user goes back to administrator role. In case the user did not have previous role it goes to volunteer.
![Screen Shot 2023-03-03 at 4 25 26 PM (2)](https://user-images.githubusercontent.com/83235086/222844534-e6e02cb3-0747-4496-8c93-68a9657773f7.png)
![IMG_2378](https://user-images.githubusercontent.com/83235086/222847537-e9a9d787-984f-4a92-a78a-f9b6bb3e495d.jpg)

agujjar2206 commented:
Hello @jinchaof, I tested this PR with the following steps:

1. Used the provided login as I didn't have an owner user.
2. Created an "Akshay Test Role" and assigned it to a new user.
3. Checked the new user's profile to ensure that the "Akshay Test Role" is displayed.
4. Deleted the "Akshay Test Role" in permission management.
5. Checked the new user's profile again and saw that the role status has changed to volunteer status as the default.
6. Logged in as the new user and verified that the status is at volunteer.
7. Created two new users and assigned them the "Akshay Test Role".
8. Repeated steps 4-6 and saw that both users' role status defaults to volunteer.
9. Verified that the PR worked as intended.
10. Noted that the deleted "Akshay Test Role" is still displayed in user management, as expected.

Screenshots:
<img width="1440" alt="285-1" src="https://user-images.githubusercontent.com/94432002/222873227-1e6d711c-6ab6-4f8c-9785-c167102eab37.png">
<img width="1439" alt="285-2" src="https://user-images.githubusercontent.com/94432002/222873235-30b22e01-1684-43fb-9244-2a0f95d0f982.png">
<img width="1432" alt="285-3" src="https://user-images.githubusercontent.com/94432002/222873242-fab5b9bb-9840-43cb-98e7-74b65086fcd8.png">
<img width="710" alt="285-4" src="https://user-images.githubusercontent.com/94432002/222873248-f19ae459-84a2-4169-8fbe-c4641ce08e59.png">
<img width="585" alt="285-5" src="https://user-images.githubusercontent.com/94432002/222873255-52cd3fd1-47e6-41fb-92af-4b364c322bfa.png">
<img width="728" alt="285-6" src="https://user-images.githubusercontent.com/94432002/222873261-9ebb0eba-01ba-421a-8c72-c791dd64a2e5.png">

Everything worked as intended, and I didn't encounter any issues.
Thank you.

RafaelCastilhoCastro commented:
Hi guys,
It might be interesting to follow the flow suggested by @dkamghem to try to reproduce his error. Maybe thats what happened with @Lunaliar too. When possible please confirm this results and report. Thanks.

suigunfjc commented:
Hi @dkamghem, I tried your steps on my local host and I used three accounts with different original roles: volunteer, administrator, and manager. I created a new role and assigned it to these accounts then deleted the role. All of the accounts were set to volunteer as expected. In case I missed some steps to find the bug, could you please make a video of your operations so I can find the bug? Thanks!

angelinatr15 commented:
Hi @suigunfjc, I retested this branch using different spellings. I was not able to delete some of the roles I created. After selecting delete the page sends me to an error page just as it did @NyahJ. As mentioned above this is a bug that'll have to be addressed in a new PR. 

Hi @RafaelCastilhoCastro I followed the the steps that @dkamghem and @Lunaliar used to test this branch, but did not see the bug. The first user was reassigned from Admin to the new role. The second user was created and assigned to the new role only. Once I deleted the new role both users defaulted to volunteer role. Please let me know if I missed anything. 

Unable to Delete:
![Screenshot 2023-03-04 at 3 33 23 PM](https://user-images.githubusercontent.com/97488347/222933457-7fb59fb1-3257-480d-a480-9596ac1f877f.png)

Error Page:
![Screenshot 2023-03-04 at 3 27 34 PM](https://user-images.githubusercontent.com/97488347/222933439-9c0bc0db-7031-4cb6-880d-fcae767603f7.png)

Followed Steps:

https://user-images.githubusercontent.com/97488347/222934573-806bb020-e6b3-47d7-b350-81a76f9e871c.mov


Additional Testing:

https://user-images.githubusercontent.com/97488347/222934589-741eb4de-f3e8-45c1-bb7e-290fc888065c.mov

kangjay88 reviewed (commented):
Hi @suigunfjc, 
I reviewed the PR by first pulling and updating my local branch (npm build, npm install). The code worked as intended and I did not run into any bugs. Before testing this branch, I did install Redis on my Windows computer, as I was getting the "Error: connect ECONNREFUSED 127.0.0.1:6379 error" on my backend server, which resolved after installing/starting redis server. 
https://user-images.githubusercontent.com/108297706/222945952-af6187d4-1c56-4c41-9f5b-e6165744373c.mp4

NyahJ commented:
> 

Hi Jinchao,

Sounds great, thank you!

NyahJ commented:
> Hi @suigunfjc,
> 
> Most of the function performed well, the added and deleted new roles would affect users selection range in their profile page. However, there is a small issue about the role names: I used four test roles: 1. TestWithNumber 2.test role 1 (name with number and black space) 3. Test03/02/2023 (name with numbers and "/") 4. TestWith!? (name with "?" and "!") The test result for names with characters and numbers only worked perfectly, both of adding and deleting. **But after we added the role names with punctuation successfully, we were not able to delete them, because the webpage appeared error/blank (result was different by different marks) so that we could not make it to delete those kinds of roles.** I have attached a video for your reference, could you pls take a look at it? Thanks!

NyahJ commented:
> > Hi @suigunfjc,
> > Most of the function performed well, the added and deleted new roles would affect users selection range in their profile page. However, there is a small issue about the role names: I used four test roles: 1. TestWithNumber 2.test role 1 (name with number and black space) 3. Test03/02/2023 (name with numbers and "/") 4. TestWith!? (name with "?" and "!") The test result for names with characters and numbers only worked perfectly, both of adding and deleting. **But after we added the role names with punctuation successfully, we were not able to delete them, because the webpage appeared error/blank (result was different by different marks) so that we could not make it to delete those kinds of roles.** I have attached a video for your reference, could you pls take a look at it? Thanks!
> 
> Thank you @NyahJ, It is definitely a new bug and I will take a look at it. Though I think it's out of this PR's scope, we can merge this PR first and raise another PR to solve it.

Sounds great, in that case, I am gonna approve this :)

NyahJ reviewed (commented):
Was about to approve this PR, howver, it has been closed already.

dkamghem commented:
> Hi @dkamghem, I tried your steps on my local host and I used three accounts with different original roles: volunteer, administrator, and manager. I created a new role and assigned it to these accounts then deleted the role. All of the accounts were set to volunteer as expected. In case I missed some steps to find the bug, could you please make a video of your operations so I can find the bug? Thanks!

I am not sure if we all have different version of the app but i tried it again today. My default role is Administrator. It does not matter whichever role a user had, once that role is deleted the user's role goes by default to administrator role which is the first role in the list of the roles. Maybe i just do not have the most updated version of the app which I will be curious to know how to get the most updated one.

RafaelCastilhoCastro commented:
> Sorry about closing it by mistake, approved!

@NyahJ The PR should only be approved after a good number of reviews confirming all its features are working and the code is neat. The PR approval is better to be left to the leadership team to perform after the final review. When performing your reviews please pay attention to only comment it. Please pay extra attention not to close or change the PR in any other way. Thanks.

RafaelCastilhoCastro commented:
> Hi @suigunfjc, I retested this branch using different spellings. I was not able to delete some of the roles I created. After selecting delete the page sends me to an error page just as it did @NyahJ. As mentioned above this is a bug that'll have to be addressed in a new PR.
> 
> Hi @RafaelCastilhoCastro I followed the steps that @dkamghem and @Lunaliar used to test this branch, but did not see the bug. The first user was reassigned from Admin to the new role. The second user was created and assigned to the new role only. Once I deleted the new role both users defaulted to volunteer role. Please let me know if I missed anything.
> 
> Unable to Delete: ![Screenshot 2023-03-04 at 3 33 23 PM](https://user-images.githubusercontent.com/97488347/222933457-7fb59fb1-3257-480d-a480-9596ac1f877f.png)
> 
> Error Page: ![Screenshot 2023-03-04 at 3 27 34 PM](https://user-images.githubusercontent.com/97488347/222933439-9c0bc0db-7031-4cb6-880d-fcae767603f7.png)
> 
> Followed Steps:
> 
> Additional Testing:

Hi @angelinatr15,
If I understood your review correctly you are experiencing the bug. When deleting a role all users with that role should revert to Volunteer in all cases, never to any other role (@suigunfjc correct me if I'm wrong please.)

one-community commented:
Has the new bug been logged on the Bugs Doc? Can I merge this then?

angelinatr15 commented:
> Hi @angelinatr15, If I understood your review correctly you are experiencing the bug. When deleting a role all users with that role should revert to Volunteer in all cases, never to any other role (@suigunfjc correct me if I'm wrong please.)

Hi @RafaelCastilhoCastro, I am not experiencing the bug that @dkamghem and @Lunaliar is having. I think @suigunfjc is referring to a different bug.

suigunfjc commented:
@one-community Logged and I think we can merge this PR now.

one-community commented:
Thank you, merging!
